sdk/storage/azblob/blob/client.go

```diff
@@ -372,24 +369,32 @@ func (b *Client) DownloadStream(ctx context.Context, o *DownloadStreamOptions) (
 
 // DownloadBuffer downloads an Azure blob to a buffer with parallel.
 func (b *Client) DownloadBuffer(ctx context.Context, buffer []byte, o *DownloadBufferOptions) (int64, error) {
-	return b.download(ctx, shared.NewBytesWriter(buffer), o)
+	if o == nil {
+		o = &DownloadBufferOptions{}
+	}
+	return b.download(ctx, shared.NewBytesWriter(buffer), (downloadOptions)(*o))
 }
 
 // DownloadFile downloads an Azure blob to a local file.
 // The file would be truncated if the size doesn't match.
 func (b *Client) DownloadFile(ctx context.Context, file *os.File, o *DownloadFileOptions) (int64, error) {
+	if o == nil {
+		o = &DownloadFileOptions{}
+	}
+	do := (*downloadOptions)(o)
+
 	// 1. Calculate the size of the destination file
 	var size int64
 
-	count := o.Count
+	count := do.Count
 	if count == CountToEnd {
 		// Try to get Azure blob's size
-		getBlobPropertiesOptions := o.getBlobPropertiesOptions()
+		getBlobPropertiesOptions := do.getBlobPropertiesOptions()
 		props, err := b.GetProperties(ctx, getBlobPropertiesOptions)
```

Comment:
@JeffreyRichter I noticed that we make two API calls in `DownloadFile` which we try to avoid.  In this case it's to get the length of the blob.  Is there another way to achieve this or can we leave it as-is?

Reply:
Will follow up separately on this.

Reply:
To do memory-mapped file writes to a file, the file has to be sized correctly first and then memory-mapped. So, we have to get the size of the blob first, create the file & set its length, memory-map it, and then download into the MMF (in parallel). There are ways to be more clever here like do a single GET block (which includes the size and the data), use the size to create/size file & MMF it and then write the first block while now starting more concurrent GETs but this is more complicated/brittle logic for a perf improvement.

Also, for these high-level APIs that do "magic" I think of them as customers just accepting the "magic". The lower-level APIs exist that do not magic and customers can control the behavior exactly as they want.